File: ios/RNFitnessUtils.m
#import "RNFitnessUtils.h"

@implementation RNFitnessUtils

+(NSDate *)setHoursMinutesSeconds: (NSDate *)date :(NSInteger)hours :(NSInteger)minutes :(NSInteger)seconds {
    NSCalendar *cal = [NSCalendar currentCalendar];
    NSDateComponents *components = [cal components:( NSCalendarUnitDay| NSCalendarUnitMonth | NSCalendarUnitYear | NSCalendarUnitHour | NSCalendarUnitMinute | NSCalendarUnitSecond ) fromDate:date];
    [components setHour:hours];
    [components setMinute:minutes];
    [components setSecond:seconds];
    [components setMonth: components.month];
    [components setDay: components.day];
    [components setYear: components.year];
    
    return [cal dateFromComponents:components];
}

+(NSDate *)beginningOfDay:(NSDate *)date {
    return [self setHoursMinutesSeconds:date :0 :0 :0];
}

+(NSDate *)endOfDay:(NSDate *)date {
    return [self setHoursMinutesSeconds:date :23 :59 :59];
}

+(NSDate *)daysAgo: (NSDate *)date :(NSInteger)numberOfDays {
    NSDate *day = [[NSCalendar currentCalendar] dateByAddingUnit:NSCalendarUnitDay value:-numberOfDays toDate:date options:0];
    return day;
}

+(NSDate *)startOfXDaysAgo: (NSDate *)date :(NSInteger)numberOfDays {
    NSDate *day = [self daysAgo:date :numberOfDays];
    
    return [self setHoursMinutesSeconds:day :0 :0 :0];
}

+(NSString *)formatUtcIsoDateTimeString:(NSDate *)date {
    return [NSISO8601DateFormatter stringFromDate:date timeZone:[NSTimeZone timeZoneForSecondsFromGMT:0] formatOptions:NSISO8601DateFormatWithFullDate | NSISO8601DateFormatWithFullTime];
}

+(NSString *)formatUtcIsoDateString:(NSDate *)date {
    return [NSISO8601DateFormatter stringFromDate:date timeZone:[NSTimeZone timeZoneForSecondsFromGMT:0] formatOptions:NSISO8601DateFormatWithFullDate];
}

+(NSString *)formatIsoDateString:(NSDate *)date {
    return [NSISO8601DateFormatter stringFromDate:date timeZone:NSTimeZone.systemTimeZone formatOptions:NSISO8601DateFormatWithFullDate];
}


@end

File: ios/RNFitnessUtils.swift
//
//  RNFitnessUtils.swift
//  RNFitnessTracker
//
//  Created by Matas on 2022-02-10.
//  Copyright © 2022 Facebook. All rights reserved.
//

import Foundation

@objc(RNFitnessUtils)
class RNFitnessUtils: NSObject {

    private static func setHoursMinutesSeconds(date: Date, hours: NSInteger, minutes: NSInteger, seconds: NSInteger) -> Date {
        return NSCalendar.current.date(bySettingHour: hours, minute: minutes, second: seconds, of: date)!;
    }

    static func startOfDay(date: Date) -> Date {
        return self.setHoursMinutesSeconds(date: date, hours: 0, minutes: 0, seconds: 0);
    }

    static func endOfDay(date: Date) -> Date {
        return self.setHoursMinutesSeconds(date: date, hours: 23, minutes: 59, seconds: 59);
    }

    static func daysAgo(date: Date, _ numberOfDays: Int) -> Date? {
        return date.addingTimeInterval(TimeInterval(-numberOfDays*86400));
    }

    static func beginningOfDay(date: Date) -> Date {
        self.setHoursMinutesSeconds(date: date, hours: 0, minutes: 0, seconds: 0)
    }

    static func startOfXDaysAgo(date: Date, numberOfDays: Int) -> Date {
        let day: Date? = self.daysAgo(date: date, numberOfDays)

        return self.setHoursMinutesSeconds(date: day!, hours: 0, minutes: 0, seconds: 0)
    }

    static func formatIsoDateString(_ date: Date) -> String {
        let formatter: ISO8601DateFormatter = ISO8601DateFormatter.init()
        formatter.formatOptions = .withFullDate
        formatter.timeZone = TimeZone.current

        return formatter.string(from: date)
    }

    static func getDateFrom(timestamp: Int) -> Date {
        Date(timeIntervalSince1970: TimeInterval(timestamp / 1000))
    }

    static func formatUtcIsoDateTimeString(_ date: Date) -> String {
        let formatter: ISO8601DateFormatter = ISO8601DateFormatter.init()
        formatter.formatOptions = [.withFullDate, .withFullTime]
        formatter.timeZone = TimeZone.init(secondsFromGMT: 0)

        return formatter.string(from: date)
    }
}

extension Double {
    func rounded(toPlaces places: Int) -> Double {
        let divisor = pow(10.0, Double(places))
        return (self * divisor).rounded() / divisor
    }
}
